### ISSUE TITLE
Heavy string benchmark slower than 1.7

### ISSUE THREAD

**Issue by headius** [jruby/jruby#6238]
In #5857 we have been investigating a report of an application running slower, and sometimes much slower, on JRuby 9.x versus JRuby 1.7. The primary original issue in that bug turned out to be the performance of `caller` on newer JDKs, which we largely addressed and I have closed the bug.

However there's still work going on to improve the performance of apps that @rovf runs, leading up to a separate reproduction script from X and provided below:

```ruby
require 'benchmark'

REPETITIONS = (ARGV[0]||9).to_i

class Bcontainer
  ARR_SIZE=1048
  def initialize
    @bigarr=nil
  end
  def b_one(n)
    @bigarr ||= []
    arr=[]
    ARR_SIZE.times do
      s = "#{n}aaaaaaaaa,bbbbbbbbbb,ccccccccccccccccc,ddddd"
      arr << s.split(/,/).join('+')
    end
    @bigarr << arr.map(&:size).map(&:to_s).join(':')
  end
end

Benchmark.bmbm do |r|
  r.report("#{REPETITIONS} iterations on #{JRUBY_VERSION}") do
    bc = Bcontainer.new
    REPETITIONS.times do |n|
      bc.b_one(n)
    end
  end
end
```

Work to improve the performance of this benchmark has started in #6236. This bug will track further discussion of findings. See the comment thread mentioned above for additional background on this benchmark.

There's also a reduced benchmark that I have been using to work on the performance of string interpolation:

```ruby
require 'benchmark'

loop { 
  puts Benchmark.measure {
    n = 0
    while n < 100000
      n+=1
      a = []
      i = 0
      while i < 1000
        i+=1
        a << "#{n}"
      end
    end
  }
}
```

Variations on this script show that, after all changes in #6236, we are still slower than 1.7 at the actual string interpolation (the to-stringification of the Fixnum and writing that content to the String has now been optimized). However the full benchmark is now just a stone's throw away from 1.7 when taken in aggregate:

```
[] ~/projects/jruby $ jruby b_aext.rb 10000
Rehearsal ------------------------------------------------------------------------
10000 iterations on 9.3.0.0-SNAPSHOT  12.780000   0.850000  13.630000 ( 11.628093)
-------------------------------------------------------------- total: 13.630000sec

                                           user     system      total        real
10000 iterations on 9.3.0.0-SNAPSHOT  11.210000   0.100000  11.310000 ( 10.865140)

[] ~/projects/jruby $ rvm jruby-9.2.11.1 do jruby b_aext.rb 10000
Rehearsal ----------------------------------------------------------------
10000 iterations on 9.2.11.1  14.230000   0.610000  14.840000 ( 13.042141)
------------------------------------------------------ total: 14.840000sec

                                   user     system      total        real
10000 iterations on 9.2.11.1  12.690000   0.130000  12.820000 ( 12.364324)

[] ~/projects/jruby $ jruby17 b_aext.rb 10000
Rehearsal --------------------------------------------------------------
10000 iterations on 1.7.27  12.440000   0.800000  13.240000 ( 11.484000)
---------------------------------------------------- total: 13.240000sec

                                 user     system      total        real
10000 iterations on 1.7.27  10.870000   0.210000  11.080000 ( 10.592000)
```

**Comment by headius** [comment]
When the benchmark is modified to make the ARR_SIZE array static (created outside the method and reused, to eliminate the cost of string interpolation and `split`) we are comfortably faster than 1.7. This seems to indicate that `map`, `join`, and the uses here of `Symbol#to_proc` are not a problem.

**Comment by headius** [comment]
I've completed work in #6236 and it now appears that JRuby 9.3 can match or best JRuby 1.7 on the given benchmark. Hopefully this will help the app from #5857 make up the difference.

This can be marked fix once #6236 has been merged.
